docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+
+.. figure:: ../data/reference/programming_model/thread_hierarchy_coop.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside sets of different sized rectangles layered on
+        oneanother titled "Block". Each "Block" containing sets of uniform
+        rectangles layered on oneanother titled "Warp". Each of the "Warp"
+        titled rectangles filled with downward pointing arrows inside.
+
+  Cooperative group thread hierarchy.
+
+* Multi Grid is an abstraction of potentially multiple simultaneous launches of
+  the same kernel over multiple devices. Grids inside a multi device kernel
+  launch need not be of uniform size, thus allowing taking into account
+  different device capabilities and preferences.
+
+  .. deprecated:: 5.0
+
+* Same as the :ref:`inherent_thread_model` Grid entity. The ability to
+  synchronize over a grid requires the kernel to be launched using the
+  Cooperative Groups API.
+
+* The defining feature of a cluster or block cluster is that all threads in a
+  cluster will share a common set of distributed shared memory which they may
+  use to share data or synchronize with oneanother.
+
+* Same as the :ref:`inherent_thread_model` Block entity.
+
+.. note::
+
+  Explicit warp-level thread handling is absent from the Cooperative Groups API.
+  In order to exploit the known hardware SIMD width on which built-in
+  functionality translates to simpler logic, one may use the group partitioning
+  part of the API, typically but not necessarily ``tiled_partition``.
+
+Memory Model
+===============================================================================
+
+The hierarchy of threads introduced by :ref:`inherent_thread_model` is induced
+by the memory subsystem of GPUs. :numref:`memory_hierarchy` summarizes that memory namespaces and
+how they relate to the various levels of the threading model.
+
+.. _memory_hierarchy:
+
+.. figure:: ../data/reference/programming_model/memory_hierarchy.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside on the upper half a rectangle titled "Cluster".
+        Inside it are two identical rectangles titled "Block", inside them are
+        ones titled "Local" with multiple "Warp" titled rectangles. Blocks have
+        not just Local inside, but also rectangles titled "Shared". The Shared
+        rectangles of Blocks in the same Cluster are grouped together with a
+        translucent halo titled "Cluster shared". Outside the Cluster but
+        inside the Grid is a rectangle titled "Global" with three others
+        inside: "Surface", "Texture" (same color) and "Constant" (different
+        color).
+
+  Memory hierarchy.
+
+* Local or per-thread memory is read-write storage only visible to the
+  threads defining the given variables. The size of a block for a given kernel,
+  the number of concurrent warps are limited by local memory usage.
```

Comment:
```suggestion
* Local or per-thread memory is read-write storage only visible to the
  thread defining the given variables. The size of a block for a given kernel,
  and thereby the number of concurrently running warps, are limited by local memory usage.
```
Due to this explanation I'd argue, that the depiction in figure three is wrong - "Local" should not surround the warps (which probably shouldn't even be in that figure), but should only surround one single thread.

Reply:
I'll add an issue for it